### ISSUE TITLE
Replacing ofBorg with Github actions

### ISSUE THREAD

**Issue by Mic92** [NixOS/nixpkgs#355847]
This is one of the two plans to ensure we can also perform github evaluation checks in the future.

See https://discourse.nixos.org/t/infrastructure-announcement-the-future-of-ofborg-your-help-needed/56025
for more information.

To replace OfBorg’s functions with GitHub Actions the following tasks need to be implemented:

- [ ] Running evaluation checks on Nixpkgs/Manual/NixOS.
- [ ] Identifying package rebuilds and adding appropriate labels to the repository.
- [ ] (Optional) Rebuilding selected packages for Linux/macOS.

I already created a proof of concept pull request here: https://github.com/NixOS/nixpkgs/pull/352808

**Comment by Mic92** [comment]
@Bot-wxt1221 I managed to run it in 5 minutes for naive nix-env evaluation based on the default.nix entry point and 15 minutes using the same logic that ofborg uses: https://github.com/Mic92/nixpkgs/actions/workflows/eval.yml

Also this is not yet the end of the line of optimizations. We still have https://github.com/Mic92/nixpkgs/blob/main/pkgs/top-level/release-attrpaths-superset.nix to split evaluation in smaller parts that can run even in parallel.

**Comment by JohnRTitor** [comment]
https://github.com/NixOS/nixpkgs/pull/352808#issuecomment-2475893708 and

I worry that bot accounts like ryantm-r can easily hit the limit of CI. CC @ryantm

**Comment by Mic92** [comment]
@JohnRTitor

Yes it's possible:

```
name: Trigger on PR Comment

on:
  issue_comment:
    types: [created]

jobs:
  run-on-comment:
    if: github.event.issue.pull_request != null && contains(github.event.comment.body, '/build')
    runs-on: ubuntu-latest
    steps:
      - name: Check out code
        uses: actions/checkout@v3
```

**Comment by Mic92** [comment]
> [#352808 (comment)](https://github.com/NixOS/nixpkgs/pull/352808#issuecomment-2475893708) and
> 
> I worry that bot accounts like ryantm-r can easily hit the limit of CI. CC @ryantm

Well. We have to try and see. Just now it's speculation if it works or not.

**Comment by JohnRTitor** [comment]
Good to know, though huge builds like kernel and its modules, chromium and firefox will obviously not work. And we'll possibly have to setup a blacklist else even individual contributors will hit their limits.

**Comment by Mic92** [comment]
> Good to know, though huge builds like kernel and its modules, chromium and firefox will obviously not work. And we'll possibly have to setup a blacklist else even individual contributors will hit their limits.

You can run builds for 12h. Obviously we should establish some reasonable timeouts to be a good citizen in the ecosystem.

**Comment by JohnRTitor** [comment]
Since this is a pretty major change related to maintainance and everyday tasks. I am pinning this to attract more feedback.

**Comment by Mic92** [comment]
Added a ^ meeting date for this.

**Comment by ibizaman** [comment]
Maybe of interest for this issue, at least just for inspiration, but I've also (ab)used GitHub actions to build tests in my project using a dynamically generated matrix. My project uses flakes but this should be adaptable to non-flakes https://github.com/ibizaman/selfhostblocks/blob/main/.github/workflows/build.yaml
This matrix then produces a big list of jobs, one job per test https://github.com/ibizaman/selfhostblocks/actions/runs/11502502422 like so:
![image](https://github.com/user-attachments/assets/8538949f-b7f7-45e2-84ca-a5af4b84cedf)

**Comment by Mic92** [comment]
See the meeting notes for today's infra meeting where we mainly discussed the CI situation: https://github.com/NixOS/infra/blob/7688f20babbeb27a10e4d8669fffe4b0ed00e17f/docs/meeting-notes/2024-11-14.md

Here is the high-level plan:
- Infinisil wants to take a look at evaluating nixpkgs in github actions to compute the number of changed paths
- Independently we will take a look how we can build packages.
- For the beginning we will just run github actions as they are designed as a pull_request event. This is because it's the most straight forward way and we actually have not validated if we cannot just build everything fast enough without resorting to my initial strategy.

Independently from meeting we also have other discussions about how we can develop ofborg in the future. However this might not happen before February, so we need some alternative solution in the meantime if not longer.

**Comment by infinisil** [comment]
I've opened a draft PR here for evaluating Nixpkgs using GitHub Actions: https://github.com/NixOS/nixpkgs/pull/356023. For just evaluation (and those only taking 5 minutes on each arch) instead of also building, I don't think we need to do the running-on-forks dance. Building is harder to get, but it's arguably also less important (and very orthogonal to evaluation).

**Comment by adisbladis** [comment]
One important aspect that ofborg currently provides, and that this issue doesn't mention, is the performance report.
This currently works by evaluating nixpkgs twice, once before the PR and once after.

For the majority of PRs the performance report is not important, but for work on `lib` & `stdenv`, it can be _very_ important.

The report currently does not report the impact of `checkMeta`, something that has lead to a less than stellar review experience since contributors & reviewers don't actually understand the real performance impact.

**Comment by Mic92** [comment]
> One important aspect that ofborg currently provides, and that this issue doesn't mention, is the performance report. This currently works by evaluating nixpkgs twice, once before the PR and once after.
> 
> For the majority of PRs the performance report is not important, but for work on `lib` & `stdenv`, it can be _very_ important.
> 
> The report currently does not report the impact of `checkMeta`, something that has lead to a less than stellar review experience since contributors & reviewers don't actually understand the real performance impact.

Could that be another on-demand GitHub actions job?

**Comment by adisbladis** [comment]
> > One important aspect that ofborg currently provides, and that this issue doesn't mention, is the performance report. This currently works by evaluating nixpkgs twice, once before the PR and once after.
> > ...
> 
> Could that be another on-demand GitHub actions job? We could even run automatically if certain paths has been changed.

Sounds good to me.

**Comment by JohnRTitor** [comment]
> Building linux kernel is fine on Github Actions, the CPU time is sufficient, it takes less than 2 hours to build Jovian-NixOS linux kernel, and Github Actions offer max 6 hours per run.

I am concerned about building the kernel modules (both in tree and out of tree).

**Comment by Mic92** [comment]
> > Building linux kernel is fine on Github Actions, the CPU time is sufficient, it takes less than 2 hours to build Jovian-NixOS linux kernel, and Github Actions offer max 6 hours per run.
> 
> I am concerned about building the kernel modules (both in tree and out of tree).

Well. We should be quickly able to filter out and blacklist packages we don't want to build once the source of truth lives in the repository? Also we can actually stop github actions, which was not possible with ofborg builds.

**Comment by Kamillaova** [comment]
> Maybe of interest for this issue, at least just for inspiration, but I've also (ab)used GitHub actions to build tests in my project using a dynamically generated matrix. My project uses flakes but this should be adaptable to non-flakes https://github.com/ibizaman/selfhostblocks/blob/main/.github/workflows/build.yaml This matrix then produces a big list of jobs, one job per test https://github.com/ibizaman/selfhostblocks/actions/runs/11502502422 like so: ![image](https://private-user-images.githubusercontent.com/1044950/386250395-8538949f-b7f7-45e2-84ca-a5af4b84cedf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMTkwMzUsIm5iZiI6MTczMjAxODczNSwicGF0aCI6Ii8xMDQ0OTUwLzM4NjI1MDM5NS04NTM4OTQ5Zi1iN2Y3LTQ1ZTItODRjYS1hNWFmNGI4NGNlZGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTIxODU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTIzY2M4MjJkYWJiMDI5NDc0Zjk0MjgwNzA1ZmQwNDFjYmRkOTlkYmJhZGMyYWIzYWNjY2IwZWZmODIwN2IwNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.iEDxuvxhzJC-Wmqo0gOWrY6dpP6s8zOOCBMyhh-S6rE)

did you see this? https://github.com/thecaralice/flake-gha

**Comment by lucasew** [comment]
IDK how useful this may be in this endeavour but I was playing on adding check status in Nixpkgs commits. No success. It fails with error 404. If I point to my fork it works but don't appear in the PR. If a workflow in a user needs to set a status it would require some kind of proxy, that is pretty doable using something like Cloudflare Workers basically for free.

I think this behavior is expected.

![image](https://github.com/user-attachments/assets/57601924-5f7c-4b80-81ed-d6dc6230b642)

**Comment by Mic92** [comment]
Evaluation on github works on a charm! Evaluation takes 5 minutes with ~3 minutes spent on actual eval.

**Comment by infinisil** [comment]
Minor problem: https://github.com/NixOS/nixpkgs/actions/runs/11940360927/job/33282898414#step:5:334

Should be fixable by just decreasing chunkSize here though: https://github.com/NixOS/nixpkgs/blob/43c3e8ba3184cc02ee0df7c47d4dcf016de03ec7/.github/workflows/eval.yml#L92 :)

**Comment by tomodachi94** [comment]
> Identifying package rebuilds and adding appropriate labels to the repository.
> (Optional) Rebuilding selected packages for Linux/macOS.

I wonder if we can reuse some of your (@Mic92's) previous work and make calls out to `nixpkgs-review`.

**Comment by infinisil** [comment]
@tomodachi94 I can work on making those things work :)

**Comment by lucasew** [comment]
I was experimenting on evaluating the package list as a package that takes `pkgs.path`.

That way the top level eval can be run as a normal build, getting remote builder support by extension.

Getting the difference between two nixpkgs commits is as easy as building this package in both revisions then running diff in the results.

https://github.com/NixOS/nixpkgs/pull/357934

**Comment by infinisil** [comment]
Next step: https://github.com/NixOS/nixpkgs/pull/359704

**Comment by Mic92** [comment]
Github actions now show added/removed/changed packages in the overview: https://github.com/NixOS/nixpkgs/actions/runs/12106942278 thanks to @itepastra in https://github.com/NixOS/nixpkgs/pull/360339

**Comment by Mic92** [comment]
Enable testing aliases: https://github.com/NixOS/nixpkgs/pull/360242

**Comment by Mic92** [comment]
Nixos options: https://github.com/NixOS/nixpkgs/pull/360876

**Comment by sersorrel** [comment]
has it been raised somewhere already that github actions workflows won't run on PRs created by @github-actions? e.g. automated backports triggered using labels like `backport release-24.11`. whether or not automated PR contents can be trusted implicitly, still maintainers of changed should be pinged, when that's implemented with GHA.

**Comment by JohnRTitor** [comment]
We just need a GitHub app for that, https://github.com/NixOS/org/issues/38

**Comment by JohnRTitor** [comment]
Evaluates `nixpkgs-lib-tests`: https://github.com/NixOS/nixpkgs/pull/361447

**Comment by JohnRTitor** [comment]
Since we will be using `get-merge-commit.sh` a lot, let's make it a reusable workflow: https://github.com/NixOS/nixpkgs/pull/361494

**Comment by misuzu** [comment]
I have a feature-request: make a CI check that check if commit messages fit https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

**Comment by infinisil** [comment]
@misuzu Let's use this existing issue for that discussion: https://github.com/NixOS/nixpkgs/issues/281573

**Comment by tomodachi94** [comment]
This has already been suggested; there is discussion here: https://github.com/NixOS/nixpkgs/issues/281573

**Comment by infinisil** [comment]
For requesting reviews from maintainers: https://github.com/NixOS/nixpkgs/pull/366046

**Comment by Mic92** [comment]
> GA eval runs out of memory in #303849: https://github.com/NixOS/nixpkgs/actions/runs/12339733507/job/34436410147?pr=303849
> 
> It runs through locally with `nix-build ci -A eval.full ...`.
> 
> Is this another example of [#355847 (comment)](https://github.com/NixOS/nixpkgs/issues/355847#issuecomment-2489906776), so reducing the chunkSize would help?

How reproducible is this. Also how much swap was used?

**Comment by ShamrockLee** [comment]
While working on #359641, I found that the current GH actions that evaluate `pkgs` misses the additional warning messages generated after applying the feature branch.

**Comment by FliegendeWurst** [comment]
As reported in https://github.com/NixOS/nixpkgs/pull/366168#issuecomment-2551838595, both GHA eval and ofborg eval do not catch some Darwin regressions. Both eval checks succeed but nixpkgs-review fails.

**Comment by Mic92** [comment]
@FliegendeWurst is it because of unfree

**Comment by FliegendeWurst** [comment]
> @FliegendeWurst is it because of unfree

The package has `license = lib.licenses.gpl2Only;`.

**Comment by Mic92** [comment]
> > @FliegendeWurst is it because of unfree
> 
> The package has `license = lib.licenses.gpl2Only;`.

Any unfree dependencies?

**Comment by FliegendeWurst** [comment]
None: jdk21, openjfx21, maven, fetchFromGitHub, makeDesktopItem, copyDesktopItems, wrapGAppsHook3, gtk3.

**Comment by lucasew** [comment]
Not very related to ofborg itself but now we have a nixpkgs-review bot based on GitHub Actions from Brazil :brazil:.

I trigger builds using a Telegram bot command and it notifies me back when all builds are done, and, of course nixpkgs-review post the comments on PRs.

A feature that we would like is some way to add a text into the comment so we could link to the workflow logs there.

This is the repo: https://github.com/nixosbrasil/nixpkgs-reviewd

If @lucasewd comments on any PR there it is because of this.

**Comment by doronbehar** [comment]
> How reproducible is this? We could also try to increase swap if it happens rarely. But also I haven't seen this in a while. Is your pr maybe massively increasing regressing on memory usage?

Another eval failure on memory:

https://github.com/NixOS/nixpkgs/actions/runs/12468993409/job/34801318849

**Comment by doronbehar** [comment]
Oh thank you - I just say the message at the bottom stating how much memory is available and I thought it was because of this. Thank you and sorry for the noise :).

**Comment by misuzu** [comment]
The `This PR is is targeting a channel branch` action fails even when the target branch was corrected:
https://github.com/NixOS/nixpkgs/actions/runs/12534682843/job/34956016297?pr=369067

**Comment by JohnRTitor** [comment]
Looks like GitHub is caching some variables (like branch info), even in subsequent re runs.

Should be fixed by a (force) push.

We can also configure the yml to run on edited events, (which includes base branch change event, pr body updated), like https://github.com/NixOS/nixpkgs/blob/adaa9f280329b5f814e8dc83eceddd42b20f72f4/.github/workflows/nixpkgs-vet.yml#L14C1-L14C51

Wish there was just a `base_changed` event though, but that's upto GitHub https://github.com/orgs/community/discussions/35058

**Comment by GGG-KILLER** [comment]
> The `This PR is is targeting a channel branch` action fails even when the target branch was corrected:
> https://github.com/NixOS/nixpkgs/actions/runs/12534682843/job/34956016297?pr=369067

Couldn't this be solved by using GitHub branch protection rules/rulesets instead of a workflow?
Having to run a whole CI pipeline just to check that it's not - pointing to a branch seems really overkill when we could configure a rule to point to `nixos-*` and `nixpkgs-*` and then create a bot account that can bypass those rules.

**Comment by paparodeo** [comment]
https://github.com/NixOS/nixpkgs/actions/runs/12563383510/job/35025186249

out of memory

**Comment by JohnRTitor** [comment]
Some PRs eval "processing" seems to get failed, which fails subsequent Tag job.

Example: https://github.com/NixOS/nixpkgs/actions/runs/12563389748/job/35025265089?pr=369760 (https://github.com/NixOS/nixpkgs/pull/369760)

**Comment by paparodeo** [comment]
> Some PRs eval "processing" seems to get failed, which fails subsequent Tag job.
> 
> Example: https://github.com/NixOS/nixpkgs/actions/runs/12563389748/job/35025265089?pr=369760 (#369760)

looks due to the out of memory failure in https://github.com/NixOS/nixpkgs/issues/355847#issuecomment-2566688172

**Comment by JohnRTitor** [comment]
> https://github.com/NixOS/nixpkgs/actions/runs/12563383510/job/35025186249
> 
> out of memory (push to master -- subsequent count calculations will fail)

@paparodeo Perhaps we need to increase the swap size/add a swap. https://github.com/NixOS/nixpkgs/pull/356023#discussion_r1844925984

**Comment by Mic92** [comment]
@wolfgangwalther potential fix: https://github.com/NixOS/nixpkgs/pull/370186

**Comment by infinisil** [comment]
https://github.com/NixOS/nixpkgs/pull/370456

**Comment by emilazy** [comment]
The eval action doesn’t properly fail on warnings, which is a regression from ofborg; see https://github.com/NixOS/nixpkgs/issues/371223. (It’d be great to get some insight into what attribute is causing the warnings, as that was a consistent pain point of ofborg-eval, but not failing at all is the bigger issue.)

**Comment by JohnRTitor** [comment]
Another issue regarding Tag subjob of Eval failing:

https://github.com/NixOS/nixpkgs/actions/runs/12646467735/job/35237397411?pr=371701

**Comment by FliegendeWurst** [comment]
> Another issue regarding Tag subjob of Eval failing on #371701:
> 
> [NixOS/nixpkgs/actions/runs/12646467735/job/35237397411?pr=371701](https://github.com/NixOS/nixpkgs/actions/runs/12646467735/job/35237397411?pr=371701)

(It didn't fail, it was skipped.) And the reason is that regular merges by GHA are not evaluated by CI.

**Comment by JohnRTitor** [comment]
> * do the periodic merges with an app token, which should allow triggering CI and thus eval?

Sounds like the most appropriate/easiest option.

**Comment by JohnRTitor** [comment]
Created https://github.com/NixOS/nixpkgs/pull/372041

**Comment by doronbehar** [comment]
I got this failure when pushing a commit to a r-ryantm PR:

```
Fetching the repository
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=2 origin 9d4b004fbb88c54361ddebebb9dd9e89d36f5385
  Error: fatal: remote error: upload-pack: not our ref 9d4b004fbb88c54361ddebebb9dd9e89d36f5385
  The process '/usr/bin/git' failed with exit code 128
  Waiting 12 seconds before trying again
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=2 origin 9d4b004fbb88c54361ddebebb9dd9e89d36f5385
  Error: fatal: remote error: upload-pack: not our ref 9d4b004fbb88c54361ddebebb9dd9e89d36f5385
  The process '/usr/bin/git' failed with exit code 128
  Waiting 10 seconds before trying again
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=2 origin 9d4b004fbb88c54361ddebebb9dd9e89d36f5385
  Error: fatal: remote error: upload-pack: not our ref 9d4b004fbb88c54361ddebebb9dd9e89d36f5385
  Error: The process '/usr/bin/git' failed with exit code 128
```

Not bothering me too much, but thought people here would like to know...

**Comment by JohnRTitor** [comment]
That seems like a .nix issue.

Another thing, would be nice if we could get a performance report like Ofborg did.

There was already an [implementation in SQL](https://github.com/NixOS/nixpkgs/commit/7ddcf2fafdde4bc5381fd5ee35afb192aca541cd) by @paparodeo (https://github.com/NixOS/nixpkgs/pull/362844#issuecomment-2525234085), [example](https://github.com/paparodeo/nixpkgs/actions/runs/12237323935). Perhaps anyone wants to reimplement it in Nix?

**Comment by infinisil** [comment]
> Eval / Process fails with `error: stack overflow; max-call-depth exceeded` for a [treewide](https://github.com/NixOS/nixpkgs/pull/376800) in https://github.com/NixOS/nixpkgs/actions/runs/12976078847/job/36187969238 - haven't seen that before.

Addressed by https://github.com/NixOS/nixpkgs/pull/377434

**Comment by infinisil** [comment]
This needs to be fixed still: https://github.com/NixOS/nixpkgs/issues/371223

**Comment by doronbehar** [comment]
Can anyone help me please understand [this eval error](https://github.com/NixOS/nixpkgs/actions/runs/13220072711/job/36904111878?pr=380485):

![Image](https://github.com/user-attachments/assets/3e46b248-cadf-43d7-8768-36a2eae5b644)

🙏

**Comment by doronbehar** [comment]
> > Can anyone help me please understand [this eval error](https://github.com/NixOS/nixpkgs/actions/runs/13220072711/job/36904111878?pr=380485):
> 
> I think this is just GHA being very flaky lately. Sometimes jobs fail for no reason and work fine when you restart them... Not sure what the root cause is.

I restarted it and it failed exactly the same way... After a rebase and force push it got fixed. Could it be it was due to an eval error that was present in branch `master` from which my PR's branch descended?

**Comment by 7c6f434c** [comment]
One thing that also happens, is GHA being flaky in a shard-dependent way where changing the commit hash rerolls the dice; typically this case doesn't look like your example, though. But maybe they have found new ways to be flaky.

**Comment by lucasew** [comment]
What if we start using impure derivations and do all this composition system using Nix itself?

GH Actions would only setup and run Nix once.

That way we could also run more extensive tests on packages.

**Comment by pbsds** [comment]
> After a rebase and force push it got fixed.

For me "Rerun all jobs" seems to work. Could it be that one of the previous jobs fail to upload some artifact?

**Comment by booxter** [comment]
Let's say I hit the issue mentioned in [this](https://github.com/NixOS/nixpkgs/issues/355847#issuecomment-2614027773) comment. How do I restart the [failed job](https://github.com/NixOS/nixpkgs/actions/runs/13348854637/job/37282807210?pr=377863)? Do I need some additional permissions to do so? (I don't see a Re-run button.) Do I have to rebase to re-trigger?

**Comment by JohnRTitor** [comment]
For some reason, sometimes, maintainer reviews are not being requested at all.

Like in https://github.com/NixOS/nixpkgs/pull/386098, workflow run (https://github.com/NixOS/nixpkgs/actions/runs/13605481225/job/38036587122, requesting maintainers reviews silently fails.

![Image](https://github.com/user-attachments/assets/92e545b4-d660-4832-86db-551eddf14575)

**Comment by infinisil** [comment]
Just talked with @hsjobeki, who's interested in implementing the stat comparison. Here's some notes I took for that:

### Stat comparison

To do:
- Change the derivation in https://github.com/NixOS/nixpkgs/blob/master/ci/eval/compare/default.nix to output the stat diffs into $out/step-summary.md
  - Stats are in `${beforeResultDir}/stats.json` `${afterResultDir}/stats.json`

To test:
- Download an after `result` artifact from any PR (e.g. https://github.com/NixOS/nixpkgs/actions/runs/13968615281?pr=391547)
- In the same PR, "Process / Target Run ID", you can get to the eval for the base branch (e.g. https://github.com/NixOS/nixpkgs/actions/runs/13968404113)
- Download the same `result` thing there
- Create `changedpaths.json` containing just `[]`
- Run this in `./ci` to test locally:
  ```
  nix-build -A eval.compare --arg beforeResultDir ./before --arg afterResultDir ./after --arg touchedFilesJson ./changedpaths.json
  ```
- Make sure markdown gets rendered correctly

Keep in mind:
- eval summary is length-limited: https://github.com/NixOS/nixpkgs/pull/365366


Step summary will be visible here in practice: https://github.com/NixOS/nixpkgs/actions/runs/13968615281?pr=391547#summary-39105012609

**Comment by JohnRTitor** [comment]
> That's because the [comparison script](https://github.com/NixOS/nixpkgs/blob/2a9b93cff7bab4775e94acbcb665e1a29ce70d97/ci/eval/compare/maintainers.nix#L51-L72) tries to match "changed files" with derivations.

I wonder if it should try to evaluate/check meta.maintainers of changed files too.

**Comment by eljamm** [comment]
Eval fails in #396950 because of [missing githubId](https://github.com/NixOS/nixpkgs/actions/runs/14341203518/job/40200924735?pr=396950#step:9:167), but I checked the package's maintainers and it seems fine, so I'm not sure what's the issue.

**Comment by JohnRTitor** [comment]
> Just talked with [@hsjobeki](https://github.com/hsjobeki), who's interested in implementing the stat comparison

Any updates?

**Comment by infinisil** [comment]
@JohnRTitor https://github.com/NixOS/nixpkgs/pull/395113 was merged but had to be reverted due to a minor problem

**Comment by doronbehar** [comment]
Does anyone have an idea why GH actions don't detect the changed packages in this `perlPackages` PR?

- https://github.com/NixOS/nixpkgs/pull/406708

Clearly `perlPackages` is supposed to be recursed into..

https://github.com/NixOS/nixpkgs/blob/48f276a14f7ae44816ea5226d203787644fa1aeb/pkgs/top-level/all-packages.nix#L10097-L10101

It is also apparent when running `nis search`.

**Comment by FliegendeWurst** [comment]
> Does anyone have an idea why GH actions don't detect the changed packages in this `perlPackages` PR?

I'd guess that only changing `meta.broken` doesn't count as change. (Since it doesn't affect the store path)

**Comment by jopejoe1** [comment]
Also the `perlPackages` attrset  gets removed during ci and hydra eval, so there is not even a chance for ci to check it, but the versioned `perl*Packages` should still be checked.

https://github.com/NixOS/nixpkgs/blob/3c5ac8ecfa2c25e33cbf31c6ddd0f5a9f2acec13/pkgs/top-level/release.nix#L406

**Comment by doronbehar** [comment]
> Also the `perlPackages` attrset gets removed during ci and hydra eval, so there is not even a chance for ci to check it, but the versioned `perl*Packages` should still be checked.
> 
> [nixpkgs/pkgs/top-level/release.nix](https://github.com/NixOS/nixpkgs/blob/3c5ac8ecfa2c25e33cbf31c6ddd0f5a9f2acec13/pkgs/top-level/release.nix#L406)
> 
> Line 406 in [3c5ac8e](/NixOS/nixpkgs/commit/3c5ac8ecfa2c25e33cbf31c6ddd0f5a9f2acec13)
>  perlPackages = { };

I too was thinking myself in that direction at first. It is true that for (legacy) ofborg building packages according to the Git commits prefixes, the Git commits probably need to be changed to `perl540Packages`.

However, I didn't intend to talk about this detail, rather I was only thinking of the PR labels and the changed attributes CI artifacts - which shouldn't be affected by the commit messages, only the commit contents.

> I'd guess that only changing meta.broken doesn't count as change. (Since it doesn't affect the store path)

That seems more like it. Did Ofborg's CI behave the same?

**Comment by winterqt** [comment]
Adding https://github.com/NixOS/nixpkgs/pull/409482#issuecomment-2902173831 to the list so it's not forgotten. I guess I'll fix it at some point.

**Comment by BirdeeHub** [comment]
> It's because GitHub is having issues right now.

I feel like that is kinda what is being discussed? About the strain on the system?

**Comment by wolfgangwalther** [comment]
I went through the whole issue and checked all the comments. We have the following open items:

> [ ]  Eval NixOS options

Tracked in #416241 now.

> [ ]  (Optional) Rebuilding selected packages for Linux/macOS

Since ofborg is still running, we won't need to focus on actually *building* in GHA right now.

> Adding [#409482 (comment)](https://github.com/NixOS/nixpkgs/pull/409482#issuecomment-2902173831) to the list so it's not forgotten. I guess I'll fix it at some point.

Tracked in #416240 now.

---

Since we have achieved our goal of replacing many parts of OfBorg with GHA and this ticket has become mostly a dump for many unrelated CI issues, I'll close it now.

Please create separate issues for anything else that comes up with CI.

**Comment by MattSturgeon** [comment]
> * (Optional) Rebuilding selected packages for Linux/macOS.

This idea is somewhat resurrected in #506749
